File: bindings/dart/lib/src/smb.dart
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// SMB protocol types for proven-servers.

/// Command matching the Idris2 ABI tags.
enum Command {
  negotiate(0),
  sessionSetup(1),
  logoff(2),
  treeConnect(3),
  treeDisconnect(4),
  create(5),
  close(6),
  read(7),
  write(8),
  lock(9),
  ioctl(10),
  cancel(11),
  queryDirectory(12),
  changeNotify(13),
  queryInfo(14),
  setInfo(15);

  const Command(this.tag);
  final int tag;

  static Command? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// Dialect matching the Idris2 ABI tags.
enum Dialect {
  smb2_0_2(0),
  smb2_1(1),
  smb3_0(2),
  smb3_0_2(3),
  smb3_1_1(4);

  const Dialect(this.tag);
  final int tag;

  static Dialect? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// ShareType matching the Idris2 ABI tags.
enum ShareType {
  disk(0),
  pipe(1),
  print(2);

  const ShareType(this.tag);
  final int tag;

  static ShareType? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// SessionState matching the Idris2 ABI tags.
enum SessionState {
  idle(0),
  negotiated(1),
  authenticated(2),
  treeConnected(3),
  fileOpen(4),
  disconnecting(5);

  const SessionState(this.tag);
  final int tag;

  static SessionState? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

File: bindings/javascript/src/smb.js
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) Jonathan D.A. Jewell <j.d.a.jewell@open.ac.uk>
//
// SMB protocol types for proven-servers.

/** Command matching the Idris2 ABI tags. */
export const Command = Object.freeze({
  NEGOTIATE: 0,
  SESSION_SETUP: 1,
  LOGOFF: 2,
  TREE_CONNECT: 3,
  TREE_DISCONNECT: 4,
  CREATE: 5,
  CLOSE: 6,
  READ: 7,
  WRITE: 8,
  LOCK: 9,
  IOCTL: 10,
  CANCEL: 11,
  QUERY_DIRECTORY: 12,
  CHANGE_NOTIFY: 13,
  QUERY_INFO: 14,
  SET_INFO: 15,
});

/** Dialect matching the Idris2 ABI tags. */
export const Dialect = Object.freeze({
  SMB2_0_2: 0,
  SMB2_1: 1,
  SMB3_0: 2,
  SMB3_0_2: 3,
  SMB3_1_1: 4,
});

/** ShareType matching the Idris2 ABI tags. */
export const ShareType = Object.freeze({
  DISK: 0,
  PIPE: 1,
  PRINT: 2,
});

/** SessionState matching the Idris2 ABI tags. */
export const SessionState = Object.freeze({
  IDLE: 0,
  NEGOTIATED: 1,
  AUTHENTICATED: 2,
  TREE_CONNECTED: 3,
  FILE_OPEN: 4,
  DISCONNECTING: 5,
});
